besu/src/main/java/org/hyperledger/besu/cli/subcommands/storage/RebuildBonsaiStateTrieSubCommand.java

```diff
@@ -0,0 +1,360 @@
+  @CommandLine.Spec
+  private CommandLine.Model.CommandSpec spec;
+
+  /** Default Constructor. */
+  public RebuildBonsaiStateTrieSubCommand() {}
+
+  @Override
+  public void run() {
+    try (final BesuController controller = createController()) {
+      var storageConfig = controller.getDataStorageConfiguration();
+
+      if (!storageConfig.getDataStorageFormat().equals(DataStorageFormat.BONSAI)) {
+        LOG.error("Data Storage format is not BONSAI.  Refusing to rebuild state trie.");
+        System.exit(-1);
+      }
+
+      var worldStateStorage =
+          (BonsaiWorldStateKeyValueStorage)
+              controller
+                  .getStorageProvider()
+                  .createWorldStateStorage(controller.getDataStorageConfiguration());
+
+      if (!worldStateStorage.getFlatDbMode().equals(FlatDbMode.FULL)) {
+        LOG.error("Database is not fully flattened. Refusing to rebuild state trie.");
+        System.exit(-1);
+      }
+
+      BlockHashAndStateRoot blockHashAndStateRoot = BlockHashAndStateRoot.create(overrideHashes);
+
+      if (blockHashAndStateRoot == null) {
+        final BlockHeader headHeader =
+            controller.getProtocolContext().getBlockchain().getChainHeadHeader();
+
+        worldStateStorage
+            .getWorldStateRootHash()
+            // we want state root hash to either be empty or same the same as chain head
+            .filter(root -> !root.equals(headHeader.getStateRoot()))
+            .ifPresent(
+                foundRoot -> {
+                  LOG.error(
+                      "Chain head {} does not match state root {}.  Refusing to rebuild state trie.",
+                      headHeader.getStateRoot(),
+                      foundRoot);
+                  System.exit(-1);
+                });
+        blockHashAndStateRoot =
+            new BlockHashAndStateRoot(headHeader.getBlockHash(), headHeader.getStateRoot());
+      }
+      verifyAndRebuild(blockHashAndStateRoot, worldStateStorage);
+    }
+  }
+
+  /**
+   * entry point for testing, verify the block hash and state root and rebuild the trie.
+   *
+   * @param blockHashAndStateRoot record tuple for block hash and state root.
+   * @param worldStateStorage bonsai worldstate storage.
+   */
+  @VisibleForTesting
+  protected void verifyAndRebuild(
+      final BlockHashAndStateRoot blockHashAndStateRoot,
+      final BonsaiWorldStateKeyValueStorage worldStateStorage) {
+    // rebuild trie:
+    var newHash = rebuildTrie(worldStateStorage);
+
+    // write state root and block hash from the header:
+    if (!blockHashAndStateRoot.stateRoot().equals(newHash)) {
+      LOG.error(
+          "Catastrophic: calculated state root {} after state rebuild, was expecting {}.",
+          newHash,
+          blockHashAndStateRoot.stateRoot());
+      if (overrideHashes == null) {
+        LOG.error(
+            "Refusing to write mismatched block hash and state root.  Node needs manual intervention.");
```

Comment:
what manual intervention would be required at this point? would we doc this subcommand